frida_tools/tracer.py

```diff
@@ -54,6 +54,7 @@ def process_builder_arg(option, opt_str, value, parser, method, **kwargs):
                     type='string', action='callback', callback=process_builder_arg, callback_args=(pb.include_debug_symbol,))
             parser.add_option("-q", "--quiet", help="do not format output messages", action='store_true', default=False)
             parser.add_option("-d", "--decorate", help="Add module name to generated onEnter log statement", action='store_true', default=False)
+            parser.add_option("-u", "--userdata", help="Store user-defined data in state.userdata variable", metavar="STRING", type='string', default='')
```

Comment:
```suggestion
            parser.add_option("-P", "--parameters", help="Store user-defined parameters in global parameters variable", metavar="JSON", type='string', default='')
```

For consistency with [frida-inject](https://github.com/frida/frida-core/blob/76172f64573ef80bc1293442750b41867e3157b3/inject/inject.vala#L22) and frida-gadget.

Reply:
Referring to your previous comment (i.e., parsing state.userdata as JSON), I assume this comment backtracks somewhat and recommends having a JSON and non-JSON user-provided data.

My original thought was that passing JSON data allows passing an unlimited amount of data to the system.  The following JSON would allow the user to pass data to two different handler sub-systems and unstructured data:

```
{
    "subsys-1": {
        "parm-1": 777,
        "parm-2": "a string"
    },
    "subsys-2": {
        "parm-1": 888,
        "parm-2": "another string"
    },
    "unstructured": "non-JSON unstructured text"
}
```

So JSON data is as flexible as we might want to be.  Having said that, I have no problem with two types of user data, JSON and non-JSON.

So should I give it a try?

Reply:
I'm leaning towards enforcing the same constraint that we enforce in frida-inject and frida-gadget: that the root has to be a JSON object. How about:

`-P '{"size": 42}'`
`-Z name:Joe`

Where `-Z` just adds a string property to the `parameters` object – which is an empty object unless you also specified `-P`.

So `-P '{"size": 42}' -Z name:Joe` would produce a `parameters` object looking like this:

```js
parameters = {
  name: 'Joe',
  size: 42
}
```

Is this on crack or does it make sense? Haven't spent a lot of time thinking about this.